File: Best Time to Buy and Sell Stock using Strategy/Best Time to Buy and Sell Stock using Strategy.cpp
// Solution for Best Time to Buy and Sell Stock using Strategy in CPP

class Solution {
public:
    using ll = long long;

    ll maxProfit(vector<int>& prices, vector<int>& strategy, int k) {
        int totalDays = prices.size();

        // ---------------------------------------------------
        // STEP 1: Compute original total profit
        // ---------------------------------------------------
        ll baseProfit = 0;
        vector<ll> dailyProfit(totalDays);

        for (int day = 0; day < totalDays; day++) {
            // Profit contribution of each day
            dailyProfit[day] = 1LL * prices[day] * strategy[day];
            baseProfit += dailyProfit[day];
        }

        // ---------------------------------------------------
        // STEP 2: Sliding window to find best modification
        // ---------------------------------------------------
        ll originalWindowProfit = 0;   // Σ original profits in current window
        ll modifiedWindowProfit = 0;   // Σ profits after modification
        ll bestExtraGain = 0;          // Maximum improvement we can get

        int left = 0;

        for (int right = 0; right < totalDays; right++) {

            // Add current day's original profit into window
            originalWindowProfit += dailyProfit[right];

            // If current index belongs to second half of k-window,
            // it will become a SELL (strategy = 1)
            if (right - left + 1 > k / 2) {
                modifiedWindowProfit += prices[right];
            }

            // If window exceeds size k, shrink from the left
            if (right - left + 1 > k) {
                originalWindowProfit -= dailyProfit[left];

                // Remove price contribution from second-half mapping
                modifiedWindowProfit -= prices[left + k / 2];

                left++;
            }

            // If window size exactly equals k, evaluate gain
            if (right - left + 1 == k) {
                ll currentGain = modifiedWindowProfit - originalWindowProfit;
                bestExtraGain = max(bestExtraGain, currentGain);
            }
        }

        // ---------------------------------------------------
        // STEP 3: Final answer = base profit + best gain
        // ---------------------------------------------------
        return baseProfit + bestExtraGain;
    }
};

File: Best Time to Buy and Sell Stock using Strategy/Best Time to Buy and Sell Stock using Strategy.py
# Solution for Best Time to Buy and Sell Stock using Strategy in PY

class Solution:
    def maxProfit(self, prices, strategy, k):
        total_days = len(prices)

        # ---------------------------------------------
        # STEP 1: Compute base profit
        # ---------------------------------------------
        base_profit = 0
        daily_profit = [0] * total_days

        for i in range(total_days):
            daily_profit[i] = prices[i] * strategy[i]
            base_profit += daily_profit[i]

        # ---------------------------------------------
        # STEP 2: Sliding window
        # ---------------------------------------------
        original_window_profit = 0
        modified_window_profit = 0
        best_extra_gain = 0

        left = 0

        for right in range(total_days):
            original_window_profit += daily_profit[right]

            # Second half of window → SELL
            if right - left + 1 > k // 2:
                modified_window_profit += prices[right]

            # Shrink window if size > k
            if right - left + 1 > k:
                original_window_profit -= daily_profit[left]
                modified_window_profit -= prices[left + k // 2]
                left += 1

            # Evaluate when window size == k
            if right - left + 1 == k:
                gain = modified_window_profit - original_window_profit
                best_extra_gain = max(best_extra_gain, gain)

        # ---------------------------------------------
        # STEP 3: Final answer
        # ---------------------------------------------
        return base_profit + best_extra_gain
